providers/implementations/keymgmt/ml_kem_kmgmt.c

```diff
@@ -0,0 +1,523 @@
+                       "ML-KEM keys cannot be mutated");
+        return 0;
+    }
+    /* An unlikely failure mode is the parameter having some unexpected type */
+    if (!OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen))
+        return 0;
+
+    if (publen != v->pubkey_bytes) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+        return 0;
+    }
+
+    return ossl_ml_kem_parse_public_key(pubenc, publen, key);
+}
+
+static int ml_kem_gen_set_params(void *vgctx, const OSSL_PARAM params[])
+{
+    struct ml_kem_gen_ctx *gctx = vgctx;
+    const OSSL_PARAM *p;
+
+    if (gctx == NULL)
+        return 0;
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        OPENSSL_free(gctx->propq);
+        if ((gctx->propq = OPENSSL_strdup(p->data)) == NULL)
+            return 0;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ML_KEM_SEED);
+    if (p == NULL)
+        return 1;
+
+    /* Treat wrong data type as promised, but missing */
+    if (p->data_type != OSSL_PARAM_OCTET_STRING) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_SEED);
+        return 0;
+    }
+    if (p->data_size != ML_KEM_SEED_BYTES) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_SEED_LENGTH);
+        return 0;
+    }
+
+    gctx->seed = gctx->seedbuf;
+    memcpy(gctx->seed, p->data, ML_KEM_SEED_BYTES);
+
+    return 1;
+}
+
+static void *ml_kem_gen_init(void *provctx, int selection,
+                             const OSSL_PARAM params[], unsigned int variant)
+{
+    struct ml_kem_gen_ctx *gctx = NULL;
+
+    if (!ossl_prov_is_running()
+        || (gctx = OPENSSL_zalloc(sizeof(*gctx))) == NULL)
+        return NULL;
+
+    if (provctx != NULL)
+        gctx->libctx = PROV_LIBCTX_OF(provctx);
+    gctx->selection = selection;
+    gctx->variant = variant;
+
+    if (ml_kem_gen_set_params(gctx, params))
+        return gctx;
+
+    OPENSSL_free(gctx->propq);
+    OPENSSL_free(gctx);
+    return NULL;
+}
+
+static const OSSL_PARAM *ml_kem_gen_settable_params(ossl_unused void *vgctx,
+                                                    ossl_unused void *provctx)
+{
+    static OSSL_PARAM settable[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ML_KEM_SEED, NULL, 0),
+        OSSL_PARAM_END
+    };
+    return settable;
+}
+
+static void *ml_kem_gen(void *vgctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    struct ml_kem_gen_ctx *gctx = vgctx;
+    ML_KEM_KEY *key;
+    uint8_t *nopub = NULL;
+    uint8_t *seed = gctx->seed;
+    size_t slen = seed == NULL ? 0 : ML_KEM_SEED_BYTES;
+    int genok = 0;
+
+    if (gctx == NULL
+        || (key = ml_kem_new(gctx->libctx, gctx->propq, gctx->variant)) == NULL)
+        return NULL;
+
+    /* Actual keypair generation may optionally be deferred */
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
```

Comment:
I am not sure what this means.. Also don't remember seeing this in any other key manager.

Reply:
I seem to recall it was possible to generate a key with a selection that sets only "domain parameters", and then populate with import.

Reply:
There are no domain params.. This should just error.

Reply:
For now implemented the RSA-like workflow.

Reply:
> In the RSA case the `selection` is tested to see if it overlaps with `OSSL_KEYMGMT_SELECT_KEYPAIR`, but it could perversely be just `OSSL_KEYMGMT_SELECT_PUBLIC_KEY`. I don't think the implementation cares, and generates a keypair regardless. Does this matter?

No.  RSA key generation will always generate the numbers for the private and the public part, there's no way around that.  For ML-KEM, maybe there's a choice, but do what seems reasonable for the provider.

Reply:
There isn't really a sensible way to generate just a public key. So it is either generate a private key anyway, or reject the request to generate only a public key.
I was wondering which of the two was preferable? Perhaps rejecting public-key-only requests would catch inadvertent errors?

Reply:
> Perhaps rejecting public-key-only requests would catch inadvertent errors?

Yes please, reject what does not make sense here.